File: playwright-python/pages/cart_and_checkout_page.py
from playwright.sync_api import Page


class CartPage:
    def __init__(self, page: Page):
        self.page = page

        # TODO: Define locators for the cart page
        self.cart_items = None       # TODO: .cart_item
        self.item_names = None       # TODO: .inventory_item_name
        self.checkout_button = None  # TODO: data-test='checkout'

    def get_item_count(self) -> int:
        # TODO: Return number of items in the cart
        pass

    def get_item_names(self) -> list[str]:
        # TODO: Return list of item name strings in the cart
        pass

    def proceed_to_checkout(self):
        # TODO: Click the checkout button
        pass


class CheckoutPage:
    def __init__(self, page: Page):
        self.page = page

        # TODO: Define locators for checkout steps 1 and 2
        self.first_name_input = None   # TODO: data-test='firstName'
        self.last_name_input = None    # TODO: data-test='lastName'
        self.postal_code_input = None  # TODO: data-test='postalCode'
        self.continue_button = None    # TODO: data-test='continue'
        self.finish_button = None      # TODO: data-test='finish'
        self.complete_header = None    # TODO: .complete-header

    def fill_customer_info(self, first_name: str, last_name: str, postal_code: str):
        # TODO: Fill in the three fields
        pass

    def continue_to_overview(self):
        # TODO: Click the continue button
        pass

    def finish_order(self):
        # TODO: Click the finish button
        pass

    def get_complete_header(self) -> str:
        # TODO: Return the text of the order complete header
        pass

    def is_order_complete(self) -> bool:
        # TODO: Return True if the complete header is visible
        pass

File: playwright-typescript/pages/CartAndCheckoutPage.ts
import { Page, Locator } from "@playwright/test";

export class CartPage {
  readonly page: Page;
  readonly cartItems: Locator;
  readonly itemNames: Locator;
  readonly checkoutButton: Locator;

  constructor(page: Page) {
    this.page = page;
    this.cartItems = page.locator(".cart_item");
    this.itemNames = page.locator(".inventory_item_name");
    this.checkoutButton = page.locator("[data-test='checkout']");
  }

  async getItemCount(): Promise<number> {
    return await this.cartItems.count();
  }

  async getItemNames(): Promise<string[]> {
    return await this.itemNames.allInnerTexts();
  }

  async proceedToCheckout() {
    await this.checkoutButton.click();
  }
}

export class CheckoutPage {
  readonly page: Page;
  readonly firstNameInput: Locator;
  readonly lastNameInput: Locator;
  readonly postalCodeInput: Locator;
  readonly continueButton: Locator;
  readonly finishButton: Locator;
  readonly completeHeader: Locator;

  constructor(page: Page) {
    this.page = page;
    this.firstNameInput = page.locator("[data-test='firstName']");
    this.lastNameInput = page.locator("[data-test='lastName']");
    this.postalCodeInput = page.locator("[data-test='postalCode']");
    this.continueButton = page.locator("[data-test='continue']");
    this.finishButton = page.locator("[data-test='finish']");
    this.completeHeader = page.locator(".complete-header");
  }

  async fillCustomerInfo(firstName: string, lastName: string, postalCode: string) {
    await this.firstNameInput.fill(firstName);
    await this.lastNameInput.fill(lastName);
    await this.postalCodeInput.fill(postalCode);
  }

  async continueToOverview() {
    await this.continueButton.click();
  }

  async finishOrder() {
    await this.finishButton.click();
  }

  async getCompleteHeader(): Promise<string> {
    return await this.completeHeader.innerText();
  }

  async isOrderComplete(): Promise<boolean> {
    return await this.completeHeader.isVisible();
  }
}
